### PR TITLE
[training, fix] Support packed sequence packing with pipeline parallelism

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

Enable variable_seq_lengths when in-batch packing is used with PP > 1, and make labels/loss_mask/attention_mask optional in pack_batch_sequences to handle non-last PP stages that don't carry these tensors.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

## Summary by CodeRabbit

* **Improvements**
  * Enhanced variable-length sequence handling in distributed training configurations
  * Increased robustness of batch sequence packing to support optional input fields

### CODE REVIEW

**Review by yaoyu-33 — DISMISSED** [review]

File: src/megatron/bridge/training/config.py [hunk 1]
```diff
@@ -1940,6 +1940,17 @@ def runtime_config_update(cfg: ConfigContainer) -> None:
     # Validate configuration after all modifications
     cfg.validate()
 
+    # Enable variable_seq_lengths for in-batch packing with pipeline parallelism.
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/config.py]: why not directly put in finalize?

Comment (cuichenx) [reply src/megatron/bridge/training/config.py]: do you mean in TransformerConfig.finalize()? I guess the challenge is that it won't have access to cfg.dataset.pack_sequences_in_batch there

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2825/commits/bed561007ea8a85c5cc09551f10ac948bc45ab38

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2825/commits/e1aa0b29c641357bbca197a47082bb536edf23d4

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2825/commits/42191c8c65af586c727a7408ca5a7a9c02d17319

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2825/commits/9747c8f501d5885152f0e8fa6969b804f5f5d2f7

**Review by yaoyu-33 — APPROVED** [review]

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #2825 - [training, fix] Support packed sequence packing with pipeline parallelism
- **Issue**: PR is the fix itself (no linked issue)

### 2. Root Cause
Packed sequences with PP > 1 failed because:
1. `variable_seq_lengths` wasn't enabled when in-batch packing used with PP > 1
2. `labels/loss_mask/attention_mask` were required in `pack_batch_sequences` but non-last PP stages don't carry these tensors

### 3. Trigger Configuration
- In-batch sequence packing with PP > 1
- Non-last pipeline stages

### 4. Nature of the Bug
**Classification**: CODE BUG - Packed sequences failing with pipeline parallelism

### 5. Existing Test Coverage
**In Fix PR**: YES - 1 test file:
- `tests/unit_tests/training/test_vlm_step.py`

**In NMFW Tests**: Not specifically covering packed sequences with PP

### 6. Coverage Assessment
| Test Type | Exists | Covers Bug |
|-----------|--------|------------|
| Fix PR unit tests | YES | YES |
| NMFW regression tests | NO | N/A |

### 7. New Regression Test
**NOT NEEDED** - Fix PR includes unit tests for VLM step with packed sequences.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE - Fix PR includes unit tests for PP packed sequence support.
